x/evm/abci.go

```diff
@@ -1,18 +1,170 @@
+		return types.Command{}, false
+	}
+
+	if !common.IsHexAddress(e.ContractAddress) {
+		bk.Logger(ctx).Info(fmt.Sprintf("invalid contract address %s for chain %s", e.ContractAddress, e.DestinationChain))
+		return types.Command{}, false
+	}
+
+	coin := sdk.NewCoin(asset, sdk.Int(e.Amount))
+	fee, err := n.ComputeTransferFee(ctx, sourceChain, destinationChain, coin)
+	if err != nil {
+		bk.Logger(ctx).Info(fmt.Sprintf("failed computing transfer fee for event %s with error %s", event.GetID(), err.Error()))
+		return types.Command{}, false
+	}
+
+	if coin.IsLT(fee) {
+		bk.Logger(ctx).Info(fmt.Sprintf("amount %s less than fee %s", e.Amount.String(), fee.Amount.String()))
+		return types.Command{}, false
+	}
+
+	destinationChainID, ok := destinationCk.GetChainID(ctx)
+	if !ok {
+		panic(fmt.Errorf("could not find chain ID for '%s'", destinationChain.Name))
+	}
+
+	keyID, ok := s.GetCurrentKeyID(ctx, destinationChain, tss.SecondaryKey)
+	if !ok {
+		panic(fmt.Errorf("no secondary key for chain %s found", destinationChain.Name))
+	}
+
+	amount := e.Amount.Sub(sdk.Uint(fee.Amount))
+	cmd, err := types.CreateApproveContractCallWithMintCommand(
+		destinationChainID,
+		keyID,
+		sourceChain.Name,
+		event.TxId,
+		event.Index,
+		*e,
+		amount,
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	n.AddTransferFee(ctx, fee)
+
+	return cmd, true
+}
+
+func handleConfirmedEvents(ctx sdk.Context, bk types.BaseKeeper, n types.Nexus, s types.Signer) error {
+	for _, chain := range n.GetChains(ctx) {
+		// skip if not an evm chain
+		if !bk.HasChain(ctx, chain.Name) {
+			continue
+		}
+
+		// skip if not activated
+		if !n.IsChainActivated(ctx, chain) {
+			continue
+		}
+
+		ck := bk.ForChain(chain.Name)
+		// skip if gateway not set yet
+		if _, ok := ck.GetGatewayAddress(ctx); !ok {
+			continue
+		}
+
+		queue := ck.GetContractCallQueue(ctx)
+		// skip if confirmed event queue is empty
+		if queue.IsEmpty() {
+			continue
+		}
+
+		var event types.Event
+		for queue.Dequeue(&event) {
+			var cmd types.Command
+			var ok bool
+
+			switch event.GetEvent().(type) {
+			case *types.Event_ContractCallWithToken:
+				cmd, ok = handleContractCallWithToken(ctx, event, bk, n, s)
+			default:
+				return fmt.Errorf("unsupported event type %T", event)
+			}
+
+			switch ok {
+			case true:
+				if err := ck.EnqueueCommand(ctx, cmd); err != nil {
+					return err
+				}
+
+				if err := ck.SetEventCompleted(ctx, event.GetID()); err != nil {
+					return err
+				}
+
+				ck.Logger(ctx).Debug(fmt.Sprintf("created %s command for event", cmd.Command),
+					"chain", chain.Name,
+					"eventID", event.GetID(),
+					"commandID", cmd.ID.Hex(),
+				)
+			default:
+				if err := ck.SetEventFailed(ctx, event.GetID()); err != nil {
+					return err
+				}
+
+				ck.Logger(ctx).Debug("failed creating command for event",
+					"chain", chain.Name,
+					"eventID", event.GetID(),
+				)
+			}
```

Comment:
I get where you are coming from with this switch for a bool, but I think this is cleaner

```
        if !ok {
		if err := ck.SetEventFailed(ctx, event.GetID()); err != nil {
			return err
		}

		ck.Logger(ctx).Debug("failed creating command for event",
			"chain", chain.Name,
			"eventID", event.GetID(),
		)
		continue
	}

	if err := ck.EnqueueCommand(ctx, cmd); err != nil {
		return err
	}

	if err := ck.SetEventCompleted(ctx, event.GetID()); err != nil {
		return err
	}

	ck.Logger(ctx).Debug(fmt.Sprintf("created %s command for event", cmd.Command),
		"chain", chain.Name,
		"eventID", event.GetID(),
		"commandID", cmd.ID.Hex(),
	)
```

Reply:
Agree and I already changed it to this style in the next PR where I added support for TokenSent.